File: crates/prolly/bindings/node/examples/cookbook-lib.ts
import assert from "node:assert/strict";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { loadNative, type NativeModule, type NativeProllyEngine, type NativeTreeRecord } from "../src/native.ts";

const enc = new TextEncoder();
const dec = new TextDecoder();

const bytes = (value: string): Uint8Array => enc.encode(value);
const text = (value: Uint8Array | null | undefined): string | null =>
  value == null ? null : dec.decode(value);
const upsert = (key: string, value: string | Uint8Array) => ({
  kind: "upsert" as const,
  key: bytes(key),
  value: typeof value === "string" ? bytes(value) : value,
});
const del = (key: string) => ({ kind: "delete" as const, key: bytes(key) });
const rootHex = (tree: NativeTreeRecord): string => Buffer.from(tree.root ?? []).toString("hex");

async function withNative<T>(fn: (native: NativeModule) => T): Promise<T> {
  const native = await loadNative();
  return fn(native);
}

export async function batchBuild(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const entries = Array.from({ length: 64 }, (_, idx) => {
      const value = 64 - idx;
      return { key: bytes(`event/${value.toString().padStart(4, "0")}`), value: bytes(`payload-${value}`) };
    });
    const tree = engine.buildFromEntries(entries);
    const rows = engine.range(tree, bytes("event/"), bytes("event0"));
    const stats = engine.collectStatsJson(tree);

    assert.equal(rows.length, 64);
    assert.equal(text(rows[0].key), "event/0001");
    assert.match(stats, /num_nodes/);

    console.log(`batch_build: imported ${rows.length} events`);
  });
}

export async function localFirstState(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const main = bytes("app/demo/root/main");
    const base = engine.batch(engine.create(), [
      upsert("entity/user/001", "Ada"),
      upsert("index/user/name/Ada/001", new Uint8Array()),
    ]);
    engine.publishNamedRoot(main, base);

    const device = engine.batch(base, [
      upsert("entity/task/900", "offline draft"),
      upsert("index/task/status/open/900", new Uint8Array()),
    ]);
    const canonical = engine.put(base, bytes("entity/user/002"), bytes("Grace"));
    engine.publishNamedRoot(main, canonical);

    const current = engine.loadNamedRoot(main)!;
    const merged = engine.merge(base, current, device, "prefer_right");
    const update = engine.compareAndSwapNamedRoot(main, current, merged);

    assert.equal(update.applied, true);
    assert.equal(text(engine.get(merged, bytes("entity/user/002"))), "Grace");
    assert.equal(text(engine.get(merged, bytes("entity/task/900"))), "offline draft");

    console.log("local_first_state: merged offline branch into main");
  });
}

export async function resolver(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const base = engine.put(engine.create(), bytes("settings/theme"), bytes("light"));
    const leftDelete = engine.delete(base, bytes("settings/theme"));
    const rightUpdate = engine.put(base, bytes("settings/theme"), bytes("dark"));

    const updateWins = engine.merge(base, leftDelete, rightUpdate, "update_wins");
    const deleteWins = engine.merge(base, leftDelete, rightUpdate, "delete_wins");

    assert.equal(text(engine.get(updateWins, bytes("settings/theme"))), "dark");
    assert.equal(engine.get(deleteWins, bytes("settings/theme")), null);

    console.log("resolver: demonstrated update-wins and delete-wins policies");
  });
}

export async function crdtMerge(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const empty = engine.create();
    const baseValue = native.timestampedValueToBytes({ value: bytes("base"), timestamp: "1" });
    const leftValue = native.timestampedValueToBytes({ value: bytes("left"), timestamp: "2" });
    const rightValue = native.timestampedValueToBytes({ value: bytes("right"), timestamp: "3" });

    const base = engine.put(empty, bytes("counter/global"), baseValue);
    const left = engine.put(base, bytes("counter/global"), leftValue);
    const right = engine.put(base, bytes("counter/global"), rightValue);
    const merged = engine.crdtMerge(base, left, right, native.crdtConfigLww("update_wins"));
    const decoded = native.timestampedValueFromBytes(engine.get(merged, bytes("counter/global"))!);
    const mergedSet = native.multiValueSetMerge([bytes("candidate-b")], [bytes("candidate-a"), bytes("candidate-b")]);

    assert.equal(text(decoded.value), "right");
    assert.equal(decoded.timestamp, "3");
    assert.deepEqual(mergedSet.map(text), ["candidate-a", "candidate-b"]);

    console.log("crdt_merge: last-writer-wins and multi-value helpers passed");
  });
}

export async function conversationMemory(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const main = bytes("conversation/c42/root/main");
    const attemptName = bytes("conversation/c42/attempt/extractor/a1");

    const base = engine.put(engine.create(), bytes("conversation/c42/memory/m001"), bytes("user|likes terse summaries|0.91"));
    engine.publishNamedRoot(main, base);
    const attempt = engine.put(base, bytes("conversation/c42/memory/m002"), bytes("user|uses TypeScript|0.87"));
    engine.publishNamedRoot(attemptName, attempt);
    const canonical = engine.put(base, bytes("conversation/c42/memory/m003"), bytes("user|prefers local-first apps|0.82"));
    engine.publishNamedRoot(main, canonical);

    const merged = engine.merge(base, engine.loadNamedRoot(main)!, engine.loadNamedRoot(attemptName)!, "prefer_right");
    const update = engine.compareAndSwapNamedRoot(main, canonical, merged);

    assert.equal(update.applied, true);
    assert.equal(engine.range(merged, bytes("conversation/c42/memory/"), bytes("conversation/c42/memory0")).length, 3);

    console.log("conversation_memory: accepted extractor attempt into canonical memory");
  });
}

export async function agentEventLog(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const root = bytes("agent-log/run-7/root/events/current");
    const tree = engine.batch(engine.create(), [
      upsert("agent-log/run-7/event/1783036805000/0001", "user|Summarize the plan"),
      upsert("agent-log/run-7/event/1783036805000/0002", "tool-call|search-docs"),
      upsert("agent-log/run-7/event/1783036806000/0003", "assistant|Plan ready"),
    ]);
    engine.publishNamedRoot(root, tree);

    const page = engine.rangePage(engine.loadNamedRoot(root)!, null, null, "2");
    assert.equal(page.entries.length, 2);
    assert.ok(page.nextCursor);

    console.log(`agent_event_log: first page has ${page.entries.length} events`);
  });
}

export async function backgroundCompaction(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const events = engine.batch(
      engine.create(),
      Array.from({ length: 6 }, (_, idx) => upsert(`event/${(idx + 1).toString().padStart(4, "0")}`, `raw-event-${idx + 1}`)),
    );
    engine.publishNamedRoot(bytes("compaction/run/r7/root/events/0001"), events);

    const compacted = engine.batch(events, [
      del("event/0001"),
      del("event/0002"),
      del("event/0003"),
      del("event/0004"),
      upsert("event/0004-summary", "summary of events 1..4"),
    ]);
    engine.publishNamedRoot(bytes("compaction/run/r7/root/events/current"), compacted);

    const plan = engine.planStoreGc([events, compacted]);
    const remaining = engine.range(compacted, bytes("event/"), bytes("event0"));
    assert.equal(remaining.length, 3);
    assert.ok(Number(plan.reclaimableNodes) >= 0);

    console.log(`background_compaction: compacted log to ${remaining.length} records`);
  });
}

export async function deterministicRagSnapshot(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const indexRoot = bytes("rag/corpus/docs/root/index/current");
    const indexV1 = engine.batch(engine.create(), [
      upsert("rag/corpus/docs/chunk/doc-1/0001", "vector:v1|CrabDB stores deterministic roots"),
      upsert("rag/corpus/docs/chunk/doc-2/0001", "vector:v2|Prolly trees diff by key"),
    ]);
    engine.publishNamedRoot(indexRoot, indexV1);

    const answers = engine.put(engine.create(), bytes("rag/answer/q1"), bytes(`query:q1|snapshot:${rootHex(indexV1)}|citation:doc-1/0001`));
    engine.publishNamedRoot(bytes("rag/corpus/docs/root/answers"), answers);

    const indexV2 = engine.put(indexV1, bytes("rag/corpus/docs/chunk/doc-3/0001"), bytes("vector:v3|New content"));
    engine.publishNamedRoot(indexRoot, indexV2);

    assert.equal(engine.range(indexV1, bytes("rag/corpus/docs/chunk/"), bytes("rag/corpus/docs/chunk0")).length, 2);
    assert.equal(engine.range(engine.loadNamedRoot(indexRoot)!, bytes("rag/corpus/docs/chunk/"), bytes("rag/corpus/docs/chunk0")).length, 3);

    console.log("deterministic_rag_snapshot: replay kept original index root");
  });
}

export async function documentChunkIndex(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const blobStore = native.NativeProllyBlobStore.memory();
    const textKey = bytes("doc-index/corpus/text/parser-v1/doc-1/chunk-0001");
    const metadataKey = bytes("doc-index/corpus/parser/parser-v1/document/doc-1/chunk/000000");

    let tree = engine.create();
    tree = engine.putLargeValue(blobStore, tree, textKey, bytes("CrabDB stores large chunk text outside prolly leaves.".repeat(8)), { inlineThreshold: "32" });
    tree = engine.put(tree, metadataKey, bytes("doc-1|chunk-0001|0|384|vector-0001"));

    assert.equal(engine.range(tree, bytes("doc-index/corpus/parser/"), bytes("doc-index/corpus/parser0")).length, 1);
    assert.ok(text(engine.getLargeValue(blobStore, tree, textKey))?.startsWith("CrabDB stores"));

    console.log("document_chunk_index: metadata and blob-backed chunk text are linked");
  });
}

export async function vectorSidecar(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const sidecar = new Map([
      ["vec-1", [0.9, 0.1]],
      ["vec-2", [0.8, 0.2]],
      ["vec-stale", [1.0, 0.0]],
    ]);
    const tree = engine.batch(engine.create(), [
      upsert("vector-sidecar/corpus/docs/chunk/doc-1/0001", "vec-1|doc-1|parser-v1"),
      upsert("vector-sidecar/corpus/docs/chunk/doc-2/0001", "vec-2|doc-2|parser-v1"),
    ]);
    const allowed = new Set(
      engine
        .range(tree, bytes("vector-sidecar/corpus/docs/chunk/"), bytes("vector-sidecar/corpus/docs/chunk0"))
        .map((entry) => text(entry.value)!.split("|", 1)[0]),
    );
    const hits = [...sidecar.keys()].sort().filter((vectorId) => allowed.has(vectorId));
    assert.deepEqual(hits, ["vec-1", "vec-2"]);

    console.log(`vector_sidecar: filtered sidecar hits to ${hits.length} snapshot vectors`);
  });
}

export async function provenanceValues(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const source = bytes("CrabDB language bindings design");
    const sourceCid = Buffer.from(native.cidFromBytes(source)).toString("hex");
    const chunkCid = Buffer.from(native.cidFromBytes(source.slice(0, 16))).toString("hex");
    const tree = engine.batch(engine.create(), [
      upsert("provenance/chunk/file-1/chunk-1", `source=${sourceCid}|chunk=${chunkCid}|parser=v1`),
      upsert("provenance/claim/file-1/claim-1", "CrabDB uses Rust-backed bindings|chunk=file-1/chunk-1"),
    ]);

    const claims = engine.range(tree, bytes("provenance/claim/file-1/"), bytes("provenance/claim/file-10"));
    assert.equal(claims.length, 1);
    assert.match(text(claims[0].value)!, /Rust-backed/);

    console.log("provenance_values: claim links back to source and chunk CIDs");
  });
}

type Order = { tenant: string; id: string; status: string; cents: number };
const orderKey = (order: Order): Uint8Array => bytes(`orders/source/tenant/${order.tenant}/order/${order.id}`);
const encodeOrder = (order: Order): Uint8Array => bytes(`${order.tenant}|${order.id}|${order.status}|${order.cents}`);
const decodeOrder = (value: Uint8Array): Order => {
  const [tenant, id, status, cents] = text(value)!.split("|");
  return { tenant, id, status, cents: Number(cents) };
};
const viewKey = (tenant: string, status: string): Uint8Array => bytes(`orders/view/by-status/tenant/${tenant}/status/${status}`);

function buildRevenueView(engine: NativeProllyEngine, source: NativeTreeRecord): NativeTreeRecord {
  const totals = new Map<string, number>();
  for (const entry of engine.range(source, bytes("orders/source/"), bytes("orders/source0"))) {
    const order = decodeOrder(entry.value);
    const key = `${order.tenant}|${order.status}`;
    totals.set(key, (totals.get(key) ?? 0) + order.cents);
  }
  return engine.batch(
    engine.create(),
    [...totals.entries()].sort().map(([key, cents]) => {
      const [tenant, status] = key.split("|");
      return { kind: "upsert" as const, key: viewKey(tenant, status), value: bytes(String(cents)) };
    }),
  );
}

export async function materializedView(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const o1 = { tenant: "acme", id: "o1", status: "paid", cents: 1200 };
    const o2 = { tenant: "acme", id: "o2", status: "open", cents: 500 };
    const sourceV1 = engine.batch(engine.create(), [
      { kind: "upsert", key: orderKey(o1), value: encodeOrder(o1) },
      { kind: "upsert", key: orderKey(o2), value: encodeOrder(o2) },
    ]);
    const paidO2 = { ...o2, status: "paid" };
    const sourceV2 = engine.put(sourceV1, orderKey(paidO2), encodeOrder(paidO2));
    const viewV2 = buildRevenueView(engine, sourceV2);

    assert.equal(text(engine.get(viewV2, viewKey("acme", "paid"))), "1700");
    assert.equal(engine.get(viewV2, viewKey("acme", "open")), null);

    console.log(`materialized_view: folded ${engine.diff(sourceV1, sourceV2).length} source diff`);
  });
}

export async function filesystemSnapshot(): Promise<void> {
  await withNative((native) => {
    const engine = native.NativeProllyEngine.memory();
    const blobStore = native.NativeProllyBlobStore.memory();
    let tree = engine.create();
    for (const [path, contents] of [
      ["README.md", "# Demo\n"],
      ["src/lib.rs", "pub fn answer() -> u8 { 42 }\n"],
    ] as const) {
      tree = engine.putLargeValue(blobStore, tree, bytes(`path/${path}`), bytes(contents), { inlineThreshold: "4" });
    }
    engine.publishNamedRoot(bytes("refs/heads/main"), tree);
    const loaded = engine.loadNamedRoot(bytes("refs/heads/main"))!;
    assert.equal(text(engine.getLargeValue(blobStore, loaded, bytes("path/README.md"))), "# Demo\n");

    console.log("filesystem_snapshot: published branch with blob-backed file contents");
  });
}

export async function durableSqlite(): Promise<void> {
  await withNative((native) => {
    const dir = mkdtempSync(join(tmpdir(), "prolly-node-"));
    try {
      const engine = native.NativeProllyEngine.sqlite(join(dir, "app.prolly.sqlite"));
      const tree = engine.batch(engine.create(), [upsert("user/1", "Ada"), upsert("user/2", "Grace")]);
      engine.publishNamedRoot(bytes("users/main"), tree);
      const loaded = engine.loadNamedRoot(bytes("users/main"))!;
      assert.deepEqual(loaded.root, tree.root);
      assert.equal(text(engine.get(loaded, bytes("user/1"))), "Ada");
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }

    console.log("durable_sqlite: named root survived through SQLite store API");
  });
}

export const scenarios = {
  batch_build: batchBuild,
  local_first_state: localFirstState,
  resolver,
  crdt_merge: crdtMerge,
  conversation_memory: conversationMemory,
  agent_event_log: agentEventLog,
  background_compaction: backgroundCompaction,
  deterministic_rag_snapshot: deterministicRagSnapshot,
  document_chunk_index: documentChunkIndex,
  vector_sidecar: vectorSidecar,
  provenance_values: provenanceValues,
  materialized_view: materializedView,
  filesystem_snapshot: filesystemSnapshot,
  durable_sqlite: durableSqlite,
};

export async function runAll(): Promise<void> {
  for (const scenario of Object.values(scenarios)) {
    await scenario();
  }
}

File: crates/prolly/bindings/ruby/examples/cookbook_lib.rb
# frozen_string_literal: true

require 'tmpdir'
require 'prolly'

module Cookbook
  module_function

  def upsert(key, value)
    Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: key.b, value: value.b)
  end

  def delete(key)
    Prolly::MutationRecord.new(kind: Prolly::MutationKind::DELETE, key: key.b, value: nil)
  end

  def assert_equal(expected, actual, label = 'value')
    return if expected == actual

    raise "#{label}: expected #{expected.inspect}, got #{actual.inspect}"
  end

  def batch_build
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    entries = (1..64).to_a.reverse.map do |idx|
      Prolly::EntryRecord.new(key: format('event/%04d', idx).b, value: "payload-#{idx}".b)
    end
    tree = engine.build_from_entries(entries)
    rows = engine.range(tree, 'event/'.b, 'event0'.b)
    stats = engine.collect_stats_json(tree).json

    assert_equal 64, rows.length, 'row count'
    assert_equal 'event/0001'.b, rows.first.key, 'first key'
    raise 'stats missing num_nodes' unless stats.include?('num_nodes')

    puts "batch_build: imported #{rows.length} events"
  end

  def local_first_state
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    main = 'app/demo/root/main'.b
    base = engine.batch(
      engine.create,
      [
        upsert('entity/user/001', 'Ada'),
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'index/user/name/Ada/001'.b, value: ''.b)
      ]
    )
    engine.publish_named_root(main, base)

    device = engine.batch(
      base,
      [
        upsert('entity/task/900', 'offline draft'),
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: 'index/task/status/open/900'.b, value: ''.b)
      ]
    )
    canonical = engine.put(base, 'entity/user/002'.b, 'Grace'.b)
    engine.publish_named_root(main, canonical)

    current = engine.load_named_root(main)
    merged = engine.merge(base, current, device, 'prefer_right')
    update = engine.compare_and_swap_named_root(main, current, merged)

    assert_equal true, update.applied, 'CAS applied'
    assert_equal 'Grace'.b, engine.get(merged, 'entity/user/002'.b), 'canonical user'
    assert_equal 'offline draft'.b, engine.get(merged, 'entity/task/900'.b), 'device task'

    puts 'local_first_state: merged offline branch into main'
  end

  def resolver
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    base = engine.put(engine.create, 'settings/theme'.b, 'light'.b)
    left_delete = engine.delete(base, 'settings/theme'.b)
    right_update = engine.put(base, 'settings/theme'.b, 'dark'.b)

    update_wins = engine.merge(base, left_delete, right_update, 'update_wins')
    delete_wins = engine.merge(base, left_delete, right_update, 'delete_wins')

    assert_equal 'dark'.b, engine.get(update_wins, 'settings/theme'.b), 'update-wins setting'
    assert_equal nil, engine.get(delete_wins, 'settings/theme'.b), 'delete-wins setting'

    puts 'resolver: demonstrated update-wins and delete-wins policies'
  end

  def crdt_merge
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    base_value = Prolly.timestamped_value_to_bytes(Prolly::TimestampedValueRecord.new(value: 'base'.b, timestamp: 1))
    left_value = Prolly.timestamped_value_to_bytes(Prolly::TimestampedValueRecord.new(value: 'left'.b, timestamp: 2))
    right_value = Prolly.timestamped_value_to_bytes(Prolly::TimestampedValueRecord.new(value: 'right'.b, timestamp: 3))

    base = engine.put(engine.create, 'counter/global'.b, base_value)
    left = engine.put(base, 'counter/global'.b, left_value)
    right = engine.put(base, 'counter/global'.b, right_value)
    merged = engine.crdt_merge(base, left, right, Prolly.crdt_config_lww(Prolly::CrdtDeletePolicyKind::UPDATE_WINS))
    decoded = Prolly.timestamped_value_from_bytes(engine.get(merged, 'counter/global'.b))
    merged_set = Prolly.multi_value_set_merge(['candidate-b'.b], ['candidate-a'.b, 'candidate-b'.b])

    assert_equal 'right'.b, decoded.value, 'CRDT value'
    assert_equal 3, decoded.timestamp, 'CRDT timestamp'
    assert_equal ['candidate-a'.b, 'candidate-b'.b], merged_set, 'multi-value set'

    puts 'crdt_merge: last-writer-wins and multi-value helpers passed'
  end

  def conversation_memory
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    main = 'conversation/c42/root/main'.b
    attempt_name = 'conversation/c42/attempt/extractor/a1'.b

    base = engine.put(engine.create, 'conversation/c42/memory/m001'.b, 'user|likes terse summaries|0.91'.b)
    engine.publish_named_root(main, base)
    attempt = engine.put(base, 'conversation/c42/memory/m002'.b, 'user|uses Ruby|0.87'.b)
    engine.publish_named_root(attempt_name, attempt)
    canonical = engine.put(base, 'conversation/c42/memory/m003'.b, 'user|prefers local-first apps|0.82'.b)
    engine.publish_named_root(main, canonical)

    merged = engine.merge(base, engine.load_named_root(main), engine.load_named_root(attempt_name), 'prefer_right')
    update = engine.compare_and_swap_named_root(main, canonical, merged)
    rows = engine.range(merged, 'conversation/c42/memory/'.b, 'conversation/c42/memory0'.b)

    assert_equal true, update.applied, 'CAS applied'
    assert_equal 3, rows.length, 'memory count'

    puts 'conversation_memory: accepted extractor attempt into canonical memory'
  end

  def agent_event_log
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    root = 'agent-log/run-7/root/events/current'.b
    tree = engine.batch(
      engine.create,
      [
        upsert('agent-log/run-7/event/1783036805000/0001', 'user|Summarize the plan'),
        upsert('agent-log/run-7/event/1783036805000/0002', 'tool-call|search-docs'),
        upsert('agent-log/run-7/event/1783036806000/0003', 'assistant|Plan ready')
      ]
    )
    engine.publish_named_root(root, tree)

    page = engine.range_page(engine.load_named_root(root), nil, nil, 2)
    assert_equal 2, page.entries.length, 'page length'
    raise 'missing next cursor' unless page.next_cursor

    puts "agent_event_log: first page has #{page.entries.length} events"
  end

  def background_compaction
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    events = engine.batch(
      engine.create,
      (1..6).map { |idx| upsert(format('event/%04d', idx), "raw-event-#{idx}") }
    )
    engine.publish_named_root('compaction/run/r7/root/events/0001'.b, events)
    compacted = engine.batch(
      events,
      [
        delete('event/0001'),
        delete('event/0002'),
        delete('event/0003'),
        delete('event/0004'),
        upsert('event/0004-summary', 'summary of events 1..4')
      ]
    )
    engine.publish_named_root('compaction/run/r7/root/events/current'.b, compacted)

    plan = engine.plan_store_gc([events, compacted])
    remaining = engine.range(compacted, 'event/'.b, 'event0'.b)

    assert_equal 3, remaining.length, 'remaining records'
    raise 'invalid GC plan' if plan.reclaimable_nodes.negative?

    puts "background_compaction: compacted log to #{remaining.length} records"
  end

  def deterministic_rag_snapshot
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    index_root = 'rag/corpus/docs/root/index/current'.b
    index_v1 = engine.batch(
      engine.create,
      [
        upsert('rag/corpus/docs/chunk/doc-1/0001', 'vector:v1|CrabDB stores deterministic roots'),
        upsert('rag/corpus/docs/chunk/doc-2/0001', 'vector:v2|Prolly trees diff by key')
      ]
    )
    engine.publish_named_root(index_root, index_v1)
    answers = engine.put(engine.create, 'rag/answer/q1'.b, "query:q1|snapshot:#{index_v1.root.unpack1('H*')}|citation:doc-1/0001".b)
    engine.publish_named_root('rag/corpus/docs/root/answers'.b, answers)

    index_v2 = engine.put(index_v1, 'rag/corpus/docs/chunk/doc-3/0001'.b, 'vector:v3|New content'.b)
    engine.publish_named_root(index_root, index_v2)

    assert_equal 2, engine.range(index_v1, 'rag/corpus/docs/chunk/'.b, 'rag/corpus/docs/chunk0'.b).length, 'replay rows'
    assert_equal 3, engine.range(engine.load_named_root(index_root), 'rag/corpus/docs/chunk/'.b, 'rag/corpus/docs/chunk0'.b).length, 'current rows'

    puts 'deterministic_rag_snapshot: replay kept original index root'
  end

  def document_chunk_index
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    blob_store = Prolly::ProllyBlobStore.memory
    text_key = 'doc-index/corpus/text/parser-v1/doc-1/chunk-0001'.b
    metadata_key = 'doc-index/corpus/parser/parser-v1/document/doc-1/chunk/000000'.b

    tree = engine.put_large_value(
      blob_store,
      engine.create,
      text_key,
      ('CrabDB stores large chunk text outside prolly leaves.' * 8).b,
      Prolly::LargeValueConfigRecord.new(inline_threshold: 32)
    )
    tree = engine.put(tree, metadata_key, 'doc-1|chunk-0001|0|384|vector-0001'.b)

    metadata = engine.range(tree, 'doc-index/corpus/parser/'.b, 'doc-index/corpus/parser0'.b)
    loaded_text = engine.get_large_value(blob_store, tree, text_key)

    assert_equal 1, metadata.length, 'metadata count'
    raise 'missing chunk text' unless loaded_text.start_with?('CrabDB stores'.b)

    puts 'document_chunk_index: metadata and blob-backed chunk text are linked'
  end

  def vector_sidecar
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    sidecar = { 'vec-1' => [0.9, 0.1], 'vec-2' => [0.8, 0.2], 'vec-stale' => [1.0, 0.0] }
    tree = engine.batch(
      engine.create,
      [
        upsert('vector-sidecar/corpus/docs/chunk/doc-1/0001', 'vec-1|doc-1|parser-v1'),
        upsert('vector-sidecar/corpus/docs/chunk/doc-2/0001', 'vec-2|doc-2|parser-v1')
      ]
    )
    allowed = engine
              .range(tree, 'vector-sidecar/corpus/docs/chunk/'.b, 'vector-sidecar/corpus/docs/chunk0'.b)
              .map { |entry| entry.value.split('|'.b, 2).first }
    hits = sidecar.keys.sort.select { |vector_id| allowed.include?(vector_id.b) }

    assert_equal %w[vec-1 vec-2], hits, 'sidecar hits'

    puts "vector_sidecar: filtered sidecar hits to #{hits.length} snapshot vectors"
  end

  def provenance_values
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    source = 'CrabDB language bindings design'.b
    source_cid = Prolly.cid_from_bytes(source).unpack1('H*')
    chunk_cid = Prolly.cid_from_bytes(source[0, 16]).unpack1('H*')
    tree = engine.batch(
      engine.create,
      [
        upsert('provenance/chunk/file-1/chunk-1', "source=#{source_cid}|chunk=#{chunk_cid}|parser=v1"),
        upsert('provenance/claim/file-1/claim-1', 'CrabDB uses Rust-backed bindings|chunk=file-1/chunk-1')
      ]
    )

    claims = engine.range(tree, 'provenance/claim/file-1/'.b, 'provenance/claim/file-10'.b)
    assert_equal 1, claims.length, 'claim count'
    raise 'missing claim text' unless claims.first.value.include?('Rust-backed'.b)

    puts 'provenance_values: claim links back to source and chunk CIDs'
  end

  Order = Struct.new(:tenant, :id, :status, :cents)

  def order_key(order)
    "orders/source/tenant/#{order.tenant}/order/#{order.id}".b
  end

  def encode_order(order)
    "#{order.tenant}|#{order.id}|#{order.status}|#{order.cents}".b
  end

  def decode_order(value)
    tenant, id, status, cents = value.split('|'.b)
    Order.new(tenant, id, status, cents.to_i)
  end

  def view_key(tenant, status)
    "orders/view/by-status/tenant/#{tenant}/status/#{status}".b
  end

  def build_revenue_view(engine, source)
    totals = Hash.new(0)
    engine.range(source, 'orders/source/'.b, 'orders/source0'.b).each do |entry|
      order = decode_order(entry.value)
      totals[[order.tenant, order.status]] += order.cents
    end
    mutations = totals.sort.map do |(tenant, status), cents|
      Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: view_key(tenant, status), value: cents.to_s.b)
    end
    engine.batch(engine.create, mutations)
  end

  def materialized_view
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    o1 = Order.new('acme', 'o1', 'paid', 1200)
    o2 = Order.new('acme', 'o2', 'open', 500)
    source_v1 = engine.batch(
      engine.create,
      [
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: order_key(o1), value: encode_order(o1)),
        Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: order_key(o2), value: encode_order(o2))
      ]
    )
    paid_o2 = Order.new('acme', 'o2', 'paid', 500)
    source_v2 = engine.put(source_v1, order_key(paid_o2), encode_order(paid_o2))
    view_v2 = build_revenue_view(engine, source_v2)

    assert_equal '1700'.b, engine.get(view_v2, view_key('acme', 'paid')), 'paid revenue'
    assert_equal nil, engine.get(view_v2, view_key('acme', 'open')), 'open revenue'

    puts "materialized_view: folded #{engine.diff(source_v1, source_v2).length} source diff"
  end

  def filesystem_snapshot
    engine = Prolly::ProllyEngine.memory(Prolly.default_config)
    blob_store = Prolly::ProllyBlobStore.memory
    tree = engine.create
    { 'README.md' => "# Demo\n", 'src/lib.rs' => "pub fn answer() -> u8 { 42 }\n" }.each do |path, contents|
      tree = engine.put_large_value(
        blob_store,
        tree,
        "path/#{path}".b,
        contents.b,
        Prolly::LargeValueConfigRecord.new(inline_threshold: 4)
      )
    end
    engine.publish_named_root('refs/heads/main'.b, tree)
    loaded = engine.load_named_root('refs/heads/main'.b)
    assert_equal "# Demo\n".b, engine.get_large_value(blob_store, loaded, 'path/README.md'.b), 'README.md'

    puts 'filesystem_snapshot: published branch with blob-backed file contents'
  end

  def durable_sqlite
    Dir.mktmpdir('prolly-ruby-') do |dir|
      engine = Prolly::ProllyEngine.sqlite(File.join(dir, 'app.prolly.sqlite'), Prolly.default_config)
      tree = engine.batch(engine.create, [upsert('user/1', 'Ada'), upsert('user/2', 'Grace')])
      engine.publish_named_root('users/main'.b, tree)
      loaded = engine.load_named_root('users/main'.b)

      assert_equal tree, loaded, 'loaded SQLite root'
      assert_equal 'Ada'.b, engine.get(loaded, 'user/1'.b), 'sqlite user'
    end
    puts 'durable_sqlite: named root survived through SQLite store API'
  end

  SCENARIOS = [
    method(:batch_build),
    method(:local_first_state),
    method(:resolver),
    method(:crdt_merge),
    method(:conversation_memory),
    method(:agent_event_log),
    method(:background_compaction),
    method(:deterministic_rag_snapshot),
    method(:document_chunk_index),
    method(:vector_sidecar),
    method(:provenance_values),
    method(:materialized_view),
    method(:filesystem_snapshot),
    method(:durable_sqlite)
  ].freeze

  def run_all
    SCENARIOS.each(&:call)
  end
end
